### PR TITLE
Add support for stack unwind on success. Depends on PR #88 in uBPF

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Add support for stack unwind on success. 
Depends on: https://github.com/iovisor/ubpf/pull/88

Signed-off-by: Alan Jowett <alanjo@microsoft.com>

### CODE REVIEW

File: libs/service/api_service.cpp [hunk 1]
```diff
@@ -349,8 +354,13 @@ ebpf_verify_and_load_program(
...
 
-            for (uint32_t helper_id = 0; helper_id < helper_id_adddress.size(); helper_id++) {
-                if (ubpf_register(vm, helper_id, nullptr, reinterpret_cast<void*>(helper_id_adddress[helper_id])) < 0) {
+            for (uint32_t helper_id = 0; helper_id < helper_id_adddress_and_flags.size(); helper_id++) {
+                if (ubpf_register(
+                        vm,
+                        helper_id,
+                        nullptr,
+                        reinterpret_cast<void*>(std::get<0>(helper_id_adddress_and_flags[helper_id])),
+                        std::get<1>(helper_id_adddress_and_flags[helper_id])) < 0) {
```
Comment (shankarseal) [inline libs/service/api_service.cpp:363]: Since only the tail call is supposed to have this UNWIND_ON_SUCCESS requirement, and its a general function (implemented in EC itself), why not make that service and program modules aware of the speciality of this helper function and let it hardcode the flag in call to ubpf_register if the helper-id == tail-call? This way you wouldnt have to change the rest of the modules and save some code.

File: libs/execution_context/ebpf_program.c [hunk 2]
```diff
@@ -437,29 +438,39 @@ ebpf_program_load_byte_code(ebpf_program_t* program, ebpf_instruction_t* instruc
...
-        uint64_t out_value;
-        int ret = (uint32_t)(ubpf_exec(program->code_or_vm.vm, context, 1024, &out_value));
-        if (ret < 0) {
-            *result = ret;
+    for (count = 0; count < EBPF_MAX_TAIL_CALL_DEPTH; count++) {
+        uint64_t unwind_context = 0;
+        if (current_program->parameters.code_type == EBPF_CODE_NATIVE) {
+            ebpf_program_entry_point_t function_pointer;
+            function_pointer = (ebpf_program_entry_point_t)(current_program->code_or_vm.code.code_pointer);
+            *result = (function_pointer)(context, 1024, &unwind_context);
```
Comment (shankarseal) [inline libs/execution_context/ebpf_program.c:453]: These 2 hidden parameters will be passed using some registry. How will you ensure that no JIT compiled code uses/writes into these registry? Or are you going to make changes in the JIT compiler to always push them on stack and then make the tail call implementation aware of this hidden pointer variable?

I didnt see you modifying the tail call prototype BTW - would that be part of another change?

**Review by dthaler — CHANGES_REQUESTED** [review]
I don't think you need to get the flags along with the address, since the service already has a way to get the flags.  Specifically, in the service, you can call get_helper_prototype_windows(index) and look in the EbpfHelperPrototype to get the flags.  Then you don't need a separate copy of the flags in the kernel that have to be kept consistent with the EbpfHelperPrototype.

File: libs/execution_context/ebpf_program.c (hunk 2 above)
Comment (Alan-Jowett) [reply libs/execution_context/ebpf_program.c:453]: Yes, the Windows x64 calling convention involves passing the first parameters via registers (RCX, RDX, R8, and R9). So, functions that don't need these registers ignore them. Functions that trigger unwind write to the address contained in R8 on unwinding.

JIT compiler was modified to always store the parameter registers on the stack at known locations so that JIT code can access them later.

File: include/ebpf_helpers.h [hunk 3]
```diff
@@ -6,6 +6,8 @@
 
 #include "ebpf_structs.h"
 
+#define EBPF_MAX_TAIL_CALL_DEPTH 32
```
Comment (dthaler) [inline include/ebpf_helpers.h:9]:
```suggestion
#define MAX_TAIL_CALL_CNT 32
```
use standard define name per bpf_tail_call man page

**Comment by Alan-Jowett** [comment]
> I don't think you need to get the flags along with the address, since the service already has a way to get the flags. Specifically, in the service, you can call get_helper_prototype_windows(index) and look in the EbpfHelperPrototype to get the flags. Then you don't need a separate copy of the flags in the kernel that have to be kept consistent with the EbpfHelperPrototype.

We need this information in the kernel when we invoke the interpreter (see _ebpf_program_register_helpers). These are flags that only uBPF cares about (the verifier doesn't care about them), so mixing them with the verifier structures seems wrong.

File: libs/execution_context/ebpf_program.c [hunk 4]
```diff
@@ -437,29 +439,39 @@ ebpf_program_load_byte_code(ebpf_program_t* program, ebpf_instruction_t* instruc
...
-    if (program->parameters.code_type == EBPF_CODE_NATIVE) {
-        ebpf_program_entry_point_t function_pointer;
-        function_pointer = (ebpf_program_entry_point_t)(program->code_or_vm.code.code_pointer);
-        *result = (function_pointer)(context);
-    } else {
-        uint64_t out_value;
-        int ret = (uint32_t)(ubpf_exec(program->code_or_vm.vm, context, 1024, &out_value));
-        if (ret < 0) {
-            *result = ret;
+    for (count = 0; count < MAX_TAIL_CALL_CNT; count++) {
```
Comment (Alan-Jowett) [inline libs/execution_context/ebpf_program.c:449]: Should return an error if count == MAX_TAIL_CALL_CNT

File: include/ebpf_program_types.h [hunk 5]
```diff
@@ -48,6 +48,7 @@ typedef struct _ebpf_helper_function_addresses
 {
     uint32_t helper_function_count;
     MIDL([size_is(helper_function_count)]) uint64_t* helper_function_address;
+    MIDL([size_is(helper_function_count)]) uint64_t* helper_function_flags;
```
Comment (dthaler) [inline include/ebpf_program_types.h:51]: I still don't follow why this change is needed, since the service already gets the helper function return type and you can get the flag from that.

File: libs/execution_context/ebpf_program.h [hunk 6]
```diff
@@ -172,6 +190,17 @@ extern "C"
     void
     ebpf_program_detach_link(_Inout_ ebpf_program_t* program, _Inout_ ebpf_link_t* link);
 
+    /**
+     * @brief Store the pointer to the program to execute on tail call.
+     * 
+     * @param next_program Next program to execute.
```
Comment (dthaler) [inline libs/execution_context/ebpf_program.h:196]:
```suggestion
     * @param[in] next_program Next program to execute.
```

File: libs/platform/ebpf_state.h [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+     * @brief Uninitialize the eBPF state tracking module.
+     *
+     */
+    void
+    ebpf_state_terminate();
+
+    /**
+     * @brief Allocate a new index in the state tracker.
+     * 
+     * @param new_index Pointer to memory that contains the index on success.
```
Comment (dthaler) [inline libs/platform/ebpf_state.h:33]:
```suggestion
     * @param[out] new_index Pointer to memory that contains the index on success.
```

File: libs/platform/ebpf_state.h [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+     * @retval EBPF_SUCCESS The operation was successful.
+     * @retval EBPF_NO_MEMORY Unable to allocate resources for this
+     *  operation.
+     */
+    ebpf_result_t
+    ebpf_state_allocate_index(size_t* new_index);
```
Comment (dthaler) [inline libs/platform/ebpf_state.h:39]:
```suggestion
    ebpf_state_allocate_index(_Out_ size_t* new_index);
```

File: libs/platform/ebpf_state.h [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+    /**
+     * @brief Store a value in the state tracker.
+     * 
+     * @param index Assigned for storing state.
+     * @param value Value to be stored.
```
Comment (dthaler) [inline libs/platform/ebpf_state.h:45]:
```suggestion
     * @param[in] index Assigned for storing state.
     * @param[in] value Value to be stored.
```

File: libs/platform/ebpf_state.h [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+     *  operation.
+     */
+    ebpf_result_t
+    ebpf_state_store(size_t index, uintptr_t value);
+
+    /**
+     * @brief Load a value in the state tracker.
+     * 
+     * @param index Assigned for storing state.
+     * @param value Value to be loaded.
```
Comment (dthaler) [inline libs/platform/ebpf_state.h:57]:
```suggestion
     * @param[in] index Assigned for storing state.
     * @param value Value to be loaded.
```
is value in or out?

File: libs/platform/ebpf_state.h [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+     * @retval EBPF_SUCCESS The operation was successful.
+     * @retval EBPF_NO_MEMORY Unable to allocate resources for this
+     *  operation.
+     */
+    ebpf_result_t
+    ebpf_state_load(size_t index, uintptr_t* value);
```
Comment (dthaler) [inline libs/platform/ebpf_state.h:63]: add SAL to value

File: libs/execution_context/ebpf_program.h [hunk 12]
```diff
@@ -148,12 +165,13 @@ extern "C"
...
      * @param[out] address Address of the helper function.
+     * @param[out] flags Flags that indicate behavior of helper function.
      * @retval EBPF_SUCCESS The operation was successful.
      * @retval EBPF_INVALID_ARGUMENT The helper_function_id is not valid.
      */
     ebpf_result_t
-    ebpf_program_get_helper_function_address(
-        _In_ const ebpf_program_t* program, const uint32_t helper_function_id, uint64_t* address);
+    ebpf_program_get_helper_function_address_and_flags(
+        _In_ const ebpf_program_t* program, const uint32_t helper_function_id, uint64_t* address, uint64_t* flags);
```
Comment (dthaler) [inline libs/execution_context/ebpf_program.h:174]:
```suggestion
        _In_ const ebpf_program_t* program, const uint32_t helper_function_id, uint64_t* address, _Out_ uint64_t* flags);
```

File: libs/execution_context/ebpf_program.h [hunk 13]
```diff
@@ -173,6 +191,17 @@ extern "C"
...
+    /**
+     * @brief Store the pointer to the program to execute on tail call.
+     * 
+     * @param next_program Next program to execute.
+     * @retval EBPF_SUCCESS The operation was successful.
+     * @retval EBPF_INVALID_ARGUMENT Internal error.
+     * @retval EBPF_NO_MORE_TAIL_CALLS Program has executed to many tail calls.
+     */
+    ebpf_result_t
+    ebpf_program_set_tail_call(_In_ ebpf_program_t* next_program);
```
Comment (dthaler) [inline libs/execution_context/ebpf_program.h:203]:
```suggestion
    ebpf_program_set_tail_call(_In_ const ebpf_program_t* next_program);
```

File: libs/platform/ebpf_state.c [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+    if (_ebpf_state_next_index >= EBPF_MAX_STATE_ENTRIES) {
+        return EBPF_NO_MEMORY;
+    }
+
+    *new_index = ebpf_interlocked_increment_int64(&_ebpf_state_next_index) - 1;
+    return EBPF_SUCCESS;
+}
+
+ebpf_result_t
+_ebpf_state_get_entry(ebpf_state_entry_t** entry)
```
Comment (dthaler) [inline libs/platform/ebpf_state.c:79]:
```suggestion
_ebpf_state_get_entry(_Outptr_ ebpf_state_entry_t** entry)
```

File: libs/service/api_service.cpp [hunk 15]
```diff
@@ -86,6 +91,12 @@ _build_helper_id_to_address_map(
         }
         instruction.imm = helper_id_mapping[instruction.imm];
     }
+    for (auto& [old_helper_id, new_helper_id] : helper_id_mapping) {
+        if (get_helper_prototype_windows(old_helper_id).return_type != EBPF_RETURN_TYPE_VOID)
```
Comment (shankarseal) [inline libs/service/api_service.cpp:95]: is this the criteria for detecting tail call helper?

Comment (shankarseal) [reply libs/service/api_service.cpp:95]: I dont understand. The prototype for tail call is:
`
     "bpf_tail_call",
     EBPF_RETURN_TYPE_INTEGER,
     {EBPF_ARGUMENT_TYPE_PTR_TO_CTX, EBPF_ARGUMENT_TYPE_PTR_TO_MAP, EBPF_ARGUMENT_TYPE_ANYTHING}
`

**Review by dthaler — DISMISSED** [review]

**Comment by dthaler** [comment]
This PR addresses more of #344

File: libs/api/libbpf_map.cpp [hunk 16]
```diff
@@ -189,3 +189,9 @@ bpf_map_update_elem(int fd, const void* key, const void* value, uint64_t flags)
 {
     return libbpf_err(ebpf_map_update_element(fd, key, value, flags));
 }
+
+int
+bpf_map_lookup_elem(int fd, const void* key, void* value)
+{
+    return libbpf_err(ebpf_map_lookup_element(fd, key, value));
```
Comment (dthaler) [inline libs/api/libbpf_map.cpp:196]:
```suggestion
    return libbpf_err(-ebpf_map_lookup_element(fd, key, value));
```

The other lines you copied from were wrong too, and are fixed in PR #374

File: libs/api/libbpf_map.cpp [hunk 17]
```diff
@@ -189,3 +189,9 @@ bpf_map_update_elem(int fd, const void* key, const void* value, uint64_t flags)
 {
     return libbpf_err(ebpf_map_update_element(fd, key, value, flags));
 }
+
+int
+bpf_map_lookup_elem(int fd, const void* key, void* value)
```
Comment (dthaler) [inline libs/api/libbpf_map.cpp:194]: Need to add this API to ebpfapi\Source.def (looks like bpf_map_update_elem is missing from there too)

**Review by dthaler — APPROVED** [review]

File: libs/api/libbpf_map.cpp [hunk 18]
```diff
@@ -189,3 +189,9 @@ bpf_map_update_elem(int fd, const void* key, const void* value, uint64_t flags)
...
+{
+    return libbpf_err(-ebpf_map_lookup_element(fd, key, value));
```
[the remark this answers is not in the document]
Comment (dthaler) [reply libs/api/libbpf_map.cpp:196]: Filed issue #376

**Comment by dthaler** [comment]
@Alan-Jowett please rebase and resolve conflicts.  Is this still blocked on anything else?

**Comment by Alan-Jowett** [comment]
Yes, still waiting for https://github.com/iovisor/ubpf/pull/88.

Currently, the ubpf submodule ref points to a feature branch commit in the  https://github.com/iovisor/ubpf.git repo, not the master branch.

**Review by dthaler — APPROVED** [review]
